### ISSUE TITLE
3.2-alpha2 test_bio_addr fails on NonStop

### ISSUE THREAD

**Issue by rsbeckerca** [openssl/openssl#22218]
The `61-test_bio_addr` fails on NonStop x86 with reseeding not being detected. We are using the hardware randomizer on x86. The hardware randomizer has worked since 3.0, but this looks like it might be a new situation.

Assistance requested to localize what is or is not happening.

```
61-test_bio_addr.t ...................... 
# The results of this test will end up in test-runs/test_bio_addr
1..1
    # Subtest: ../../test/bio_addr_test
    1..1
        # Subtest: test_bio_addr_copy_dup
        1..6
        ok 1 - iteration 1
        ok 2 - iteration 2
        ok 3 - iteration 3
        ok 4 - iteration 4
        # ERROR: (bool) 'bio_addr_is_eq(src, dst) == true' failed @ /home/jenkinsbuild/.jenkins/workspace/OpenSSL-3.2_Monitor/test/bio_addr_test.c:145
        # false
        # OPENSSL_TEST_RAND_SEED=1695973622
        not ok 5 - iteration 5
        # ERROR: (bool) 'bio_addr_is_eq(src, dst) == true' failed @ /home/jenkinsbuild/.jenkins/workspace/OpenSSL-3.2_Monitor/test/bio_addr_test.c:145
        # false
        # OPENSSL_TEST_RAND_SEED=1695973622
        not ok 6 - iteration 6
    # OPENSSL_TEST_RAND_SEED=1695973622
    not ok 1 - test_bio_addr_copy_dup
../../util/wrap.pl ../../test/bio_addr_test => 255
not ok 1 - running bio_addr_test

#   Failed test 'running bio_addr_test'
#   at test/recipes/61-test_bio_addr.t line 20.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
```

**Configuration:**
```
Command line (with current working directory = .):

    perl ./Configure nonstop-nsx_64 --with-rand-seed=rdcpu enable-fips

Perl information:

    perl
    5.30.3 for NSV-D-nonstop_kernel

Enabled features:

    acvp-tests
    apps
    argon2
    aria
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    docs
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    ecx
    egd
    engine
    err
    filenames
    fips
    fips-securitychecks
    gost
    http
    idea
    legacy
    loadereng
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    quic
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm2-precomp
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    tests
    tls
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    afalgeng            [not-linux]      OPENSSL_NO_AFALGENG
    asan                [default]        OPENSSL_NO_ASAN
    asm                 [no asm_arch]    OPENSSL_NO_ASM
    brotli              [default]        OPENSSL_NO_BROTLI
    brotli-dynamic      [default]        OPENSSL_NO_BROTLI_DYNAMIC
    buildtest-c++       [default]
    winstore            [not-windows]    OPENSSL_NO_WINSTORE
    crypto-mdebug       [default]        OPENSSL_NO_CRYPTO_MDEBUG
    default-thread-pool [cascade]        OPENSSL_NO_DEFAULT_THREAD_POOL
    devcryptoeng        [default]        OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    external-tests      [default]        OPENSSL_NO_EXTERNAL_TESTS
    fuzz-afl            [default]        OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]        OPENSSL_NO_KTLS
    makedepend          [unavailable]
    md2                 [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]        OPENSSL_NO_MSAN
    rc5                 [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]        OPENSSL_NO_SCTP
    tfo                 [default]        OPENSSL_NO_TFO
    thread-pool         [cascade]        OPENSSL_NO_THREAD_POOL
    threads             [config]
    trace               [default]        OPENSSL_NO_TRACE
    ubsan               [default]        OPENSSL_NO_UBSAN
    unit-test           [default]        OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch] OPENSSL_NO_UPLINK
    weak-ssl-ciphers    [default]        OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]        OPENSSL_NO_ZLIB
    zlib-dynamic        [default]        OPENSSL_NO_ZLIB_DYNAMIC
    zstd                [default]        OPENSSL_NO_ZSTD
    zstd-dynamic        [default]        OPENSSL_NO_ZSTD_DYNAMIC
    ssl3                [default]        OPENSSL_NO_SSL3
    ssl3-method         [default]        OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "qc",
    CC => "c99",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    bn_ops => "SIXTY_FOUR_BIT",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-g -O2 -Wextensions -Wnowarn=203,220,272,734,770,1506 -Wbuild_neutral_library -Wverbose -Wtarget=tns/x -Wsystype=oss -Wlp64 -WIEEE_float",
    cppflags => "",
    defines => [ "OPENSSL_BUILDING_OPENSSL", "OPENSSL_VPROC=\$(OPENSSL_VPROC)", "_XOPEN_SOURCE", "_XOPEN_SOURCE_EXTENDED=1", "_TANDEM_SOURCE", "__NSK_OPTIONAL_TYPES__", "B_ENDIAN" ],
    disable => [ "threads" ],
    dso_scheme => "DLFCN",
    enable => [ "egd" ],
    ex_libs => "-lrld",
    includes => [  ],
    lflags => "-Wxld=\"-set systype oss\" -Wxld=\"-set data_model lp64\" -Wxld=\"-set floattype ieee_float\"",
    lib_cflags => "",
    lib_cppflags => "",
    lib_defines => [  ],
    module_cflags => "",
    module_cxxflags => undef,
    module_ldflags => "-Wshared",
    multibin => "64",
    multilib => "64",
    perl => "/usr/bin/perl",
    perl_platform => "Unix",
    shared_argfileflag => "-Wxld_obey=",
    shared_cflag => "",
    shared_defflag => "-Wxld_obey=",
    shared_defines => [  ],
    shared_extension => ".so",
    shared_ldflag => "-Wshared",
    shared_rcflag => "",
    shared_target => "nonstop-shared",
    sys_id => "TANDEM",
    thread_defines => [  ],
    thread_scheme => "(unknown)",
    unistd => "<unistd.h>",

Recorded environment:

    AR =
    ARFLAGS =
    AS =
    ASFLAGS =
    BUILDFILE =
    CC =
    CFLAGS =
    CPP =
    CPPDEFINES =
    CPPFLAGS =
    CPPINCLUDES =
    CROSS_COMPILE =
    CXX =
    CXXFLAGS =
    HASHBANGPERL =
    LD =
    LDFLAGS =
    LDLIBS =
    MT =
    MTFLAGS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL =
    RANLIB =
    RC =
    RCFLAGS =
    RM =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = ar
    ARFLAGS         = qc
    ASFLAGS         =
    CC              = c99
    CFLAGS          =
    CPPDEFINES      =
    CPPFLAGS        =
    CPPINCLUDES     =
    CXXFLAGS        =
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         =
    LDLIBS          =
    PERL            = perl
    RANLIB          = ranlib
    RC              = windres
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/common0.tmpl
    Configurations/unix-Makefile.tmpl
```

**Comment by mattcaswell** [comment]
Iterations 5 and 6 of that test are related to BIO_ADDRs that use the AF_UNIX address family. This is where we had the "fun" with BIO_ADDR previously (AF_UNIX addresses seem to be very large on NonStop).

Iteration 5 tests the BIO_ADDR_dup() function and iteration 6 tests the BIO_ADDR_copy() function. Under the covers BIO_ADDR_dup() is implemented in terms of BIO_ADDR_copy(), so its not too surprising that, if BIO_ADDR_copy() is doing something unexpected, then BIO_ADDR_dup() is too.

The test simply creates a dummy BIO_ADDR object with a fictitious address, then copies or dups it, and then compares the original BIO_ADDR with the newly copied/dup'd BIO_ADDR to make sure they are the same. It is the final step that appears to fail, i.e. it has (apparently) successfully created the original BIO_ADDR and copied/dup'd it - but when comparing the new object with the original they apparently don't contain the same address.

It would be helpful if you could add some "printf" statement, or step through with a debugger, the bio_addr_is_eq function to figure out at what point it returns 0 (indicating that it has found something that is not equal to the original):

https://github.com/openssl/openssl/blob/ad4af6dfca8344516bb658b1745a530635af9433/test/bio_addr_test.c#L81-L118

**Comment by rsbeckerca** [comment]
Thanks for the guidance @mattcaswell. I will start looking.

**Comment by mattcaswell** [comment]
@rsbeckerca - any update on this one?

**Comment by arapov** [comment]
Community Platform, is not a blocker for 3.2.

**Comment by rsbeckerca** [comment]
> @rsbeckerca - any update on this one?

I was traveling yesterday. Should have something today.

**Comment by rsbeckerca** [comment]
> Community Platform, is not a blocker for 3.2.

Based on the comments above from @mattcaswell, this is probably not a platform issue but something this platform brought to the surface, as the previous related issue. I do not agree with the assessment removal of the 3.2 milestone.

**Comment by rsbeckerca** [comment]
@mattcaswell I think your hypothesis might be correct. Here is what the first pass through shows:

```
    at /home/randall/openssl-3.2/test/bio_addr_test.c:89
*   89          if (a == b)
(xInspect 3,1199):p a
$1 = (const union bio_addr_st *) 0x10001b120
(xInspect 3,1199):p b
$2 = (const union bio_addr_st *) 0x10001b540
(xInspect 3,1199):n
*   89          if (a == b)
(xInspect 3,1199):
*   93          if (a == NULL || b == NULL)
(xInspect 3,1199):
*   96          if (BIO_ADDR_family(a) != BIO_ADDR_family(b))
(xInspect 3,1199):
*  100          if (BIO_ADDR_rawport(a) != BIO_ADDR_rawport(b))
(xInspect 3,1199):
*  100          if (BIO_ADDR_rawport(a) != BIO_ADDR_rawport(b))
(xInspect 3,1199):
*  103          if (!BIO_ADDR_rawaddress(a, NULL, &alen)
(xInspect 3,1199):
*  103          if (!BIO_ADDR_rawaddress(a, NULL, &alen)
(xInspect 3,1199):
0x0000000070004db0      *  106              return 0;
(xInspect 3,1199):p a
$3 = (const union bio_addr_st *) 0x10001b120
(xInspect 3,1199):p *a
$4 = {
  sa = {
    sa_family = 2,
    sa_data = "\003\350aaaa\000\000\000\000\000\000\000"
  },
  s_in6 = {
    sin6_family = 2,
    sin6_port = 1000,
    sin6_flowinfo = 1633771873,
    sin6_addr = {
      s6_un = {
        sa6_addr = '\000' <repeats 15 times>,
        sa6_waddr = {0, 0, 0, 0, 0, 0, 0, 0},
        sa6_laddr = {0, 0, 0, 0}
      }
    },
    sin6_scope_id = 0
  },
  s_in = {
    sin_family = 2,
    sin_port = 1000,
    sin_addr = {
      s_addr = 1633771873
    },
    sin_zero = "\000\000\000\000\000\000\000"
  },
  s_un = {
    sun_family = 2,
    sun_path = "\003\350aaaa", '\000' <repeats 1017 times>
  }
}
(xInspect 3,1199):p alen
$5 = 4
(xInspect 3,1199):p adata
$6 = {
  __ss_family = 37101,
  __ss_pad1 = "\377o\000\000\000",
  __ss_align = -7154793225914417152,
  __ss_pad2 = "\002\004\000\000\000\000\000\000\277R&x\000\000\000\000\300\355\377o\000\000\000\000\225Q&x\000\000\000\000\300\355\377o\000\000\000\000\303\264\021x\000\000\000\000\300\355\377o\000\000\000\000b\264\021x", '\000' <repeats 12 times>, "\322\003(x\000\000\000\001\000\356\377o\000\000\000\000@\265\001\000\001\000\000\000\240\002\000\b\000\000\000\000@\265\001\000\001\000\000"
}
(xInspect 3,1199):p sizeof(adata)
$8 = 128
```

**Comment by rsbeckerca** [comment]
The actual problem is here:
```
*  119          return memcmp(&adata, &bdata, alen) == 0;
```
with `adata` and `bdata` being significantly longer than alen. The two are different in content, but not at 4 bytes. That appears to be the defect.

**Comment by rsbeckerca** [comment]
Changing `alen` to `128` allowed the test to pass. The question is where should that best be set. Checking `bio_addr.c`, it is setting `*l = sizeof(ap->s_in.sin_addr);`, which is confusing when used at the length of `adata` and `bdata` in the `memcmp` above.

**Comment by t8m** [comment]
@rsbeckerca this is confusing - to pass the test the data should be identical but not necessarily the whole storage - only the part covered by the address in use is interesting, the rest might be as well uninitialized.

**Comment by t8m** [comment]
I believe you are looking at wrong iteration.

**Comment by t8m** [comment]
Could it be something like #22281?

**Comment by mattcaswell** [comment]
> Could it be something like https://github.com/openssl/openssl/pull/22281?

I doubt it - that would surely manifest as a compilation failure.

**Comment by mattcaswell** [comment]
> with adata and bdata being significantly longer than alen.

That is expected.

> The two are different in content, but not at 4 bytes

That they don't differ in the `alen` bytes is the expected outcome. In this case `memcmp` is supposed to return 0...and thus `bio_addr_is_eq` should return 1.

What we need to find is the iteration where `bio_addr_is_eq` returns 0.

I was slightly confused by your debugger output:

````
(xInspect 3,1199):
*  103          if (!BIO_ADDR_rawaddress(a, NULL, &alen)
(xInspect 3,1199):
0x0000000070004db0      *  106              return 0;
````

These line numbers don't match up to the line numbers I have for 3.2-alpha2 (off by 2). Am I reading this correctly that we are going to "return 0" on line 106 (actually line 104) after executing `BIO_ADDR_rawaddress(a, NULL, &alen)`?

**Comment by rsbeckerca** [comment]
> > with adata and bdata being significantly longer than alen.
> 
> That is expected.
> 
> > The two are different in content, but not at 4 bytes
> 
> That they don't differ in the `alen` bytes is the expected outcome. In this case `memcmp` is supposed to return 0...and thus `bio_addr_is_eq` should return 1.
> 
> What we need to find is the iteration where `bio_addr_is_eq` returns 0.
> 
> I was slightly confused by your debugger output:
> 
> ```
> (xInspect 3,1199):
> *  103          if (!BIO_ADDR_rawaddress(a, NULL, &alen)
> (xInspect 3,1199):
> 0x0000000070004db0      *  106              return 0;
> ```
> 
> These line numbers don't match up to the line numbers I have for 3.2-alpha2 (off by 2). Am I reading this correctly that we are going to "return 0" on line 106 (actually line 104) after executing `BIO_ADDR_rawaddress(a, NULL, &alen)`?

The `return 0` was not executed. It was just the debugger at the implied else (jump past the return. The `memcmp` was where the two were compared.

**Comment by rsbeckerca** [comment]
> I believe you are looking at wrong iteration.

The test failed immediately following the `memcmp` as it returned upward.

**Comment by t8m** [comment]
> The test failed immediately following the `memcmp` as it returned upward.

This does not make any sense, can you step through the code again, please? The test is checking that the copied (or duplicated) BIO_ADDR is equal to the original. I.e., the test should pass if the memcp() returns 0 - the memory is equal. And yeah the memory should be equal up to the alen bytes because following bytes are meaningless.

**Comment by t8m** [comment]
Please note that the interesting iterations are with idx == 4 and 5.

**Comment by rsbeckerca** [comment]
In iteration 5:
```
*  103          if (!BIO_ADDR_rawaddress(a, NULL, &alen)
   104                  || alen > sizeof(adata)
   105                  || !BIO_ADDR_rawaddress(a, &adata, &alen))
   106              return 0;
(xInspect 1,489):p a
$1 = (const union bio_addr_st *) 0x10001b120
(xInspect 1,489):p *a
$2 = {
  sa = {
    sa_family = 1,
    sa_data = 'a' <repeats 14 times>
  },
  s_in6 = {
    sin6_family = 1,
    sin6_port = 24929,
    sin6_flowinfo = 1633771873,
    sin6_addr = {
      s6_un = {
        sa6_addr = 'a' <repeats 16 times>,
        sa6_waddr = {24929, 24929, 24929, 24929, 24929, 24929, 24929, 24929},
        sa6_laddr = {1633771873, 1633771873, 1633771873, 1633771873}
      }
    },
    sin6_scope_id = 1633771873
  },
  s_in = {
    sin_family = 1,
    sin_port = 24929,
    sin_addr = {
      s_addr = 1633771873
Quit
(xInspect 1,489):p adata
$3 = {
  __ss_family = 37101,
  __ss_pad1 = "\377o\000\000\000",
  __ss_align = 8986107838581440512,
  __ss_pad2 = "\002\004\000\000\000\000\000\000", 'a' <repeats 24 times>, "\300\355\377o\000\000\000\000\303\264\021x\000\000\000\000\300\355\377o\000\000\000\000b\264\021x\000\000\000\000", 'a' <repeats 16 times>, "\000\356\377o\000\000\000\000@\265\001\000\001\000\000\000\240\002\000\b\000\000\000\000@\265\001\000\001\000\000"
}
(xInspect 1,489):p alen
$4 = 4295079232
```
Now, `alen` gets set to `1023` following the first entry to `BIO_ADDR_rawaddress()` at line 103. `sizeof(adata)` is 128, so the `if` fails here in the second clause of the disjunct (`||`).

My HEAD is at commit 831602922f19a8f39d0c0fae425b81e9ab402c69.

**Comment by mattcaswell** [comment]
Thanks. Please can you try out the fix in #22294?

**Comment by rsbeckerca** [comment]
> Thanks. Please can you try out the fix in #22294?

Thanks @mattcaswell 
```
61-test_bio_addr.t ..
# The results of this test will end up in test-runs/test_bio_addr
1..1
    # Subtest: ../../test/bio_addr_test
    1..1
        # Subtest: test_bio_addr_copy_dup
        1..6
        ok 1 - iteration 1
        ok 2 - iteration 2
        ok 3 - iteration 3
        ok 4 - iteration 4
        ok 5 - iteration 5
        ok 6 - iteration 6
    ok 1 - test_bio_addr_copy_dup
../../util/wrap.pl ../../test/bio_addr_test => 0
ok 1 - running bio_addr_test
ok
All tests successful.
Files=1, Tests=1,  0 wallclock secs ( 0.01 usr  0.00 sys +  0.41 cusr  0.00 csys =  0.42 CPU)
Result: PASS
make[1]: Leaving directory '/home/randall/openssl-3.2'
```

**Comment by mattcaswell** [comment]
Great! Thanks!

**Comment by rsbeckerca** [comment]
> Great! Thanks!

I was going to suggest breaking up the `if` statement to avoid side-effects but you did that. Thanks.

**Comment by nhorman** [comment]
so my fix didn't work, but something very odd is still going on here.  When working with an AF_UNIX addr, the size of the addr computed in BIO_ADDR_rawaddress should never be more than 108, as thats the size of sun_path in a sockaddr_un struct.  But here it is.  If we look at:
https://github.com/openssl/openssl/issues/22218#issuecomment-1749081838

We didn't get to see the value of the s_un part of the BIO_ADDR enumeration, but looking at the s6_un structure, you can see its all 'a's (which is in keeping with what we set all addresses to in make_dummy_addr during the test.

Given that strlen in BIO_ADDR_rawaddress is returning a huge value, I'm fairly convinced that we're somehow just failing to NULL terminate the address sometimes in BIO_ADDR_rawmake. 

One thing I note, is that in the bio addr test, specifically in make_dummy_addr we build our address based on the family name, by setting where to the address of the specific sockaddr struct we want to use, and then memcpy-ing the wherelen amount of data into it, filling it with all 'a''s, but we never initially zero the socket structure data, meaning the byte following wherelen might be NULL, or it might not.  That _should_ be ok, given that BIO_ADDR_makeraw zeros the structure that we're targeting, but somehow it seems to be happening.

@rsbeckerca can you please modify make_dummy_addr in test_bio_addr.c such that the AF_UNIX case, right before it computes the where pointer adds this line:
memset(&sa.sun.sun_path, 0, sizeof(sa.sun.sun_path)

and confirm with a debugger that that the where pointer is NULL terminated after this line:
memset(where, 'a', wherelen);

**Comment by rsbeckerca** [comment]
@nhorman Adding a memset of a potentially large buffer that will have undesirable performance impacts. The length of 1023 is not "huge", which is what was returned in the prior tests. I do not think `alen` was set when the debugger displayed its value as this was before the first call to `BIO_ADDR_rawaddress`. I am also not sure about the contention that `strlen()` is running past the end of `sun_path`. In the debug log above, `sun_path` clearly has nulls fairly close in to the front, so I do not see the `memset` causing any change in the result, however, if the buffer is uninitialized (not sure when the contents are being set in the OpenSSL code), then yes, this would be needed. @mattcaswell can you comment, please?

**Comment by nhorman** [comment]
But that's my point.  Sun_path is never more than 108 bytes, because it's statically declared in that struct.  The.fact that strlen returns anything greater than that is indicative of the fact that the string isn't null terminated, and that's a problem.  The only exception I see to that case is if the is on a nonstop platform has somehow altered the definition of struct sockaddr_un in some incompatible way

**Comment by rsbeckerca** [comment]
> But that's my point. Sun_path is never more than 108 bytes, because it's statically declared in that struct. The.fact that strlen returns anything greater than that is indicative of the fact that the string isn't null terminated, and that's a problem. The only exception I see to that case is if the is on a nonstop platform has somehow altered the definition of struct sockaddr_un in some incompatible way

`sun_path` is clearly declared as 1024 on NonStop in `sys/un.h`. I cannot find verification of the assumption that it is "never more than 108 bytes" in the POSIX specification. Please provide a reference, if able, although there is no way that that can be changed in any reasonable timeframe on the platform. From `un.h`:
```
struct  sockaddr_un {
        sa_family_t   sun_family;             /* AF_UNIX */
        char    sun_path[1024];          /* path name */
};
```
The `AF_UNIX` path allows `1024` byte socket names on the platform.

**Comment by nhorman** [comment]
Thank you, I had assumed that nonstop followed the Linux definition:
https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/un.h.html
If 108 bytes.

Having confirmed that it's 1023 bytes on nonstop, answers part of my question, though it still doesn't explain how doing a dynamic allocation resolves this issue if sockaddr_un has a 1023 byte array in it

**Comment by nhorman** [comment]
@rsbeckerca actually, would you mind sharing the definition of sockaddr_storage on nonstop please?

**Comment by rsbeckerca** [comment]
```
struct sockaddr_storage {
    sa_family_t __ss_family;             /* address family 4.3 BSD */
    char        __ss_pad1[_SS_PAD1SIZE]; /* 6 byte pad to alignment field */
    int64_t     __ss_align;              /* force structure alignment */
    char        __ss_pad2[_SS_PAD2SIZE]; /* 112 byte pad to needed size */
};
```

**Comment by nhorman** [comment]
Thats the problem!

In nonstop sockaddr_un (from your comment here https://github.com/openssl/openssl/issues/22218#issuecomment-1751826845)

Is defined to be 1024 + 16 (for sa_family_t)  = 1040 bytes, but sockaddr_storage on the same platform is only defined to be 16+6 + 64 + 112 = 198 bytes

in bio_addr_test, we test for equality of two cloned BIO_ADDR structs, which we fetch by declaring two stockaddr_storage structs (adata and bdata) on the stack.  Given that those are two automatic allocations that are adjacent, when we call BIO_ADDR_rawaddress(a, &data, &alen) in that function, we find that a->s_un.sun_path is 1024 bytes, and completely overwrite the sockaddr_storage adata struct, well into part of bdata, and beyond.  Then we call BIO_ADDR_rawaddress (a, &bdata, &blen) which does it again

Do our adata and bdata values become completely corrupt, leading to a bad result in memcmp down below in bio_addr_is_eq  

According to posix:
https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/socket.h.html

```
The <sys/socket.h> header shall define the sockaddr_storage structure. This structure shall be:

Large enough to accommodate all supported protocol-specific address structures

Aligned at an appropriate boundary so that pointers to it can be cast as pointers to protocol-specific address structures and used to access the fields of those structures without alignment problems

```

The definition of sockaddr_storage on nonstop violates that first provision, as it is not large enough to hold the currently defined sockaddr_un struct

I think the correct fix here is to modify the definition of sockaddr_storage such that:
_SS_PAD1SIZE is at least 1024 bytes
and
remove __ss_pad2

I know that kind of stinks, as modifying the OS is more difficult, but that clearly doesn't seem right, and should cause any number of other problems where sockaddr_storage is used in conjunction with AF_INET sockets on nonstop platforms.

**Comment by rsbeckerca** [comment]
I am sorry, but it is not possible to modify vendor supplied definitions - I do not work for HPE and even if I did, the answer would likely be the same. While the analysis might be correct, such a solution is not practical. Changing the operating system definition will impact millions of users who depend on OpenSSL and TCP/IP functioning properly on the platform in a manner that is compatible with prior versions of the operating system and stack. Even if a change could be made today, it would be 5-7 years before that change would take effect across the community.

A better solution is to have OpenSSL account for this situation and protect from socket definitions that are in production in the wild. As a result, I cannot accept this as a viable approach.

**Comment by nhorman** [comment]
I understand thats a slow process, but the fact remains, with the way sockaddr_storage is currently defined on nonstop, its not usable at all with AF_INET sockets there for any use case.  As noted above, we could modify our test to avoid its usage, but this really needs to get reported to HP so that they can make a proper fix, be that in a year or 10 years.

**Comment by hlandau** [comment]
This is completely noncompliant with POSIX. I'm not sure it's reasonable for us to rework this to accommodate such a broken platform. Still, we can always consider a PR, depending on how invasive it proves.

**Comment by nhorman** [comment]
@rsbeckerca you can test with this patch:
https://github.com/nhorman/openssl/tree/issue-22218

Its actually very simmilar to what @mattcaswell proposed, but it avoids heap allocation, and follows the same pattern that we use in make_dummy_addr

Though as @hlandau notes, we're really accommodating a completely broken implementation here, HP needs to do something about this, as no doubt this problem exists in many other applications that use sockaddr_storage

**Comment by rsbeckerca** [comment]
> This is completely noncompliant with POSIX. I'm not sure it's reasonable for us to rework this to accommodate such a broken platform. Still, we can always consider a PR, depending on how invasive it proves.

At worst, we can wrap the fix for NonStop in BIO_rawaddress with OPENSSL_SYS_TANDEM if a general fix is not desirable by the committers.

There are degrees of compliance for POSIX, it is not all or nothing, and categorizing NonStop as "completely noncompliant" is not constructive nor correct. The definition has been that way at least one decade, probably two, with OpenSSL correctly running on the platform for decades without encountering this issue; further, the open group specification does not explicitly discuss this point as a compliance matter - not that I can find. AF_INET does work correctly on the platform including some very esoteric use cases, so can we please avoid unjustifiable comments, please. The size of sockaddr_storage is referenced needing to be large enough to store platform content in essentially an opaque manner. Because NonStop is an MPP, with multiple CPUs and cores, multiple TCP/IP  stacks, etc., and isolated process memory, there is context that has to be maintained beyond what an SMP has. Attempting to look inside opaque data (a.k.a. sockaddr_storage) seems questionable; rather than putting the blame on the platform, which I should not have to defend in a PR discussion in the first place, although I fully agree that it is a "different" platform from others.

With all that said, it is my opinion that using `strlen()` on `sockaddr_storage` is a mistake, even comparing anything in that buffer is a red flag for me as a code reviewer. but using `memcpy()` and `memcmp()` on the entire buffer instead should be safer, under the assumption that `sun_path[1024]` contains values that make sense for the BIO address comparison, which am not sure is true or a good idea. Note that other suggestions about changing `_SS_PAD1SIZE` cannot be accepted because they do not even exist on the ia64 platform variant - but we only have 2-ish years left to worry about that before I am planning on dropping support for new ia64 builds.

**Comment by nhorman** [comment]
@rsbeckerca This is true, there are several levels of POSIX compliance (listed[ here](https://pubs.opengroup.org/onlinepubs/9699919799/)), and for each section, an implementation may or may not choose to implement it, which is fine.  An implementation may further elect to only partially implement one of those sections, for better or worse.

This isn't really about that though, NonStop chose to implement the networking api, and sepecfically chose to implement the definitions of sockaddr_[in|in6|un], and more importantly chose to implement sockaddr_storage, but did so in a way that that is simply broken.  NonStop can choose to implement something or not, but they can't implement something that violates the guarantees that the standard from which it is implemented requires of it, and expect it to work for application developers.  Thats exactly the situation we find ourselves in here at the moment.  The breaking openssl test relies on the fact that sockaddr_storage contains enough allocated data to handle any type of socket family, and it doesn't.  Thats not choosing to implement something differently, thats implementing something incorrectly. 

As for openssl working for decades, I can believe it, the test in question thats breaking was added on Sept 21 2023, and prior to that, I can see no use of sockaddr_storage in openssl at all, so I'm not suprised it worked up until recently.  That doesn't make Openssl broken here, it just means we didn't make use of a broken feature in the NonStop os until then. 

As for other users of NonStop, I'm not sure what to say. If other users used sockaddr_storage with an address that only used, say 128 or 47 or 180 bytes with a null terminated string, everything would continue to work correctly because the NonStop definition of sockaddr_storage would still encompass that data safely.  Its only if a user created a sockaddr_storage structure, cast it to a sockaddr_un and wrote a sun path that was more than 198 bytes (I think that was the max size of sockaddr_storage on NonStop) would they potentially see a problem, and only then if they were looking for memory overruns.  We only tripped over this by virtue of the fact that the bio_addr_tests used and compared two sockaddr_storage structs that were adjacent in memory.  My guess would be that Nonstop application developers that did trip over this, chose to fix the problem in their application by various workarounds, most likely because they knew that there would be significant latency in a more correct fix, but thats just conjecture on my part.

Regarding your comment that the open group doesn't discuss this as a compliance matter, I disagree.  While they don't use the word compliance specifically, they clearly indicate that sockaddr_storage [shall be large enough to accommodate all supported protocol-specific address structures](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/socket.h.html).  I don't see how failing to adhere to that requirement would not be considered a compliance violation. The standard never says that sockaddr_storage should be large enough to hold platform content in an opaque manner, it clearly indicates that it (sockaddr_sotrage) has to be both large enough to hold "all supported protocol-specific address structures", and be aligned in such a way that pointers to a sockaddr_storage can be "cast as pointers to protocol-specific address structures".  The NonStop definition fails on those, and as such isn't usable for its purpose.

I understand that NonStop is a very different environment from most normal unixes, but that doesn't excuse the platform from implementing a standard api in such a way that the guarantees that the standard makes are violated.  To do so invites exactly the issue we are seeing here.

your argument regarding the use of strlen may have merit, but thats not the issue here.  Even using memcmp on the entire buffer would be invalid in this use case, because the bottom line is that sockaddr_storage, when cast to a sockaddr_un struct (which is its purpose), is fundamentally broken here, becuase the resulting sockaddr_un pointer then believes it has access to memory well beyond what sockaddr_storage allocates.

To summarize, the purpose of sockaddr_storage is so that applications developers can create a generic socket structure, and later cast it to whatever supported type they want.  because sockaddr_storage on nonstop is 198 bytes, when we cast it to sockaddr_un (1040 bytes on nonstop), we open the possibilty that we will read or write data that was not allocated (any access between bytes 199, and 1039 will be reading/writing to memory that doesn't belong to the structure)

**Comment by levitte** [comment]
I've a possible fix in #22315.  A point is to void using `sockaddr_storage`, 'cause it's not available everywhere.  In POSIX terms, it's a fairly recent addition.  That should avoid the NonStop too.

That being said, I would urge you, @rsbeckerca, to file a report to HP.  This is very clearly a bug, and it creates all sort of impediments and having to do "special tricks" for nonstop, which isn't at all practical.
